circuit/types/group/src/helpers/from_x_coordinate.rs

```diff
@@ -26,6 +26,76 @@ impl<E: Environment> Group<E> {
+    /// in that case, the returned point is `(0, 0)`, but immaterial.
+    pub fn from_x_coordinate_flagged(x: Field<E>) -> (Boolean<E>, Self) {
+        // Obtain the A and D coefficients of the elliptic curve.
+        let a = Field::constant(console::Field::new(E::EDWARDS_A));
+        let d = Field::constant(console::Field::new(E::EDWARDS_D));
+
+        // Compute x^2.
+        let xx = &x * &x;
+
+        // Compute a * x^2 - 1.
+        let a_xx_minus_1 = a * &xx - Field::one();
+
+        // Compute d * x^2 - 1.
+        let d_xx_minus_1 = d * &xx - Field::one();
+
+        // Compute y^2 = (a * x^2 - 1) / (d * x^2 - 1), i.e. solve the curve equation for y^2.
+        let yy: Field<E> = witness!(|a_xx_minus_1, d_xx_minus_1| { a_xx_minus_1 / d_xx_minus_1 });
+        E::enforce(|| (&yy, &d_xx_minus_1, &a_xx_minus_1));
+
+        // Compute both square roots of y^2, in no specified order, with a flag saying whether y^2 is a square or not.
+        // That is, finish solving the curve equation for y.
+        // If the x-coordinate line does not intersect the elliptic curve, this returns (1, 0, 0).
+        let (yy_is_not_square, y1, y2) = yy.square_roots_flagged_nondeterministic();
+
+        // Form the two points, which are on the curve if yy_is_not_square is false.
+        // Note that the Group<E> type is not restricted to the points in the subgroup or even on the curve;
+        // it includes all possible points, i.e. all possible pairs of field elements.
+        let point1 = Self { x: x.clone(), y: y1.clone() };
+        let point2 = Self { x: x.clone(), y: y2.clone() };
+
+        // We need to check whether either of the two points is in the subgroup.
+        // There may be at most one, but in a circuit we need to always represent both computation paths.
+        // In fact, we represent this computation also when yy_is_not_square is true,
+        // in which case the results of checking whether either point is in the subgroup are meaningless,
+        // but ignored in the final selection of the results returned below.
+        // The criterion for membership in the subgroup is that
+        // multiplying the point by the subgroup order yields the zero point (0, 1).
+        // The group operation that we use here is for the type `Group<E>` of the subgroup,
+        // which as mentioned above it can be performed on points outside the subgroup as well.
+        // We turn the subgroup order into big endian bits,
+        // to get around the issue that the subgroup order is not of Scalar<E> type.
+        let order = E::ScalarField::modulus();
+        let order_bits_be = order.to_bits_be();
+        let mut order_bits_be_constants = Vec::new();
+        for bit in order_bits_be.iter() {
+            order_bits_be_constants.push(Boolean::constant(*bit));
+        }
+        let point1_times_order = order_bits_be_constants.mul(point1);
+        let point2_times_order = order_bits_be_constants.mul(point2);
```

Comment:
Can we implement a version of `Group::enforce_in_group` as `Group::is_in_group` that returns a boolean?

Reply:
Not that I know. If we impose the enforce_in_group constraints directly on the point (x, y) to check, clearly we get unsatisfiable constraints when (x, y) is not in the group. If we postulate a new/separate point (x', y') in the group (using the enforce_in_group constraints), the trouble is how to enforce the right connection between (x, y), (x', y'), and the boolean result r. Just saying that (r = 1) <==> (x = x' & y = y') is underconstrained, because even if (x, y) is in the group, I can pick a different (x', y') also in the group (because the group has more than one point) and set r = 0 (another possible solution being x = x' and y = y' and r = 1 of course).

Reply:
I'm not saying it's impossible. Maybe there's something clever, but if there is it has not occurred to any of us yet.

Reply:
Looking at `enforce_in_group` it seems like it's possible to change the `enforce_*` into `is_*`, but I don't know what the issue is.

Reply:
I looked this morning at https://github.com/AleoHQ/snarkVM/compare/fix/cast-lossy-field-to-group...feat/is-in-group (which does that) but I see an underconstraining problem similar to above. I'll DM-forward you the details.